// next-runtime/form/dom.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidFormSubmitter = exports.getFormSubmitter = exports.isInputElement = exports.isFormElement = exports.isButtonElement = exports.isHtmlElement = void 0;
function isHtmlElement(object) {
    return object != null && typeof object.tagName === 'string';
}
exports.isHtmlElement = isHtmlElement;
function isButtonElement(object) {
    return isHtmlElement(object) && object.tagName.toLowerCase() === 'button';
}
exports.isButtonElement = isButtonElement;
function isFormElement(object) {
    return isHtmlElement(object) && object.tagName.toLowerCase() === 'form';
}
exports.isFormElement = isFormElement;
function isInputElement(object) {
    return isHtmlElement(object) && object.tagName.toLowerCase() === 'input';
}
exports.isInputElement = isInputElement;
function getFormSubmitter(event) {
    return event.nativeEvent.submitter;
}
exports.getFormSubmitter = getFormSubmitter;
function isValidFormSubmitter(submitter) {
    return (isButtonElement(submitter) ||
        (isInputElement(submitter) &&
            (submitter.type === 'submit' || submitter.type === 'image')));
}
exports.isValidFormSubmitter = isValidFormSubmitter;
